### PR TITLE
Replace Bokeh Server with Tornado HTTPServer

### LINKED ISSUE #3645: Centralize HTTPServer in Server
Currently we use the HTTPServer hidden in the Bokeh server, if we create the dashboard.  We then hang various additional routes on the Tornado web application in this server for additional things, like the info pages, and prometheus.  

I suggest that we instead create a default HTTP server for the Dask server, and then attach new Tornado applications to that server as needed (like bokeh, prometheus, ...)  This might be a bit more aligned to how we actually think about things.  Additionally, if we needed to set things up like SSL, it's a bit simpler to do this within the core dask code rather than route various keywords through the bokeh code.

cc @jacobtomlinson @jcrist 

I'm making Bokeh a little better about handling this in https://github.com/bokeh/bokeh/pull/9833 , but short term we would need to bring some of that logic into our codebase (unless we wanted to depend on latest release Bokeh, which I don't think that we want to do).

### PR DESCRIPTION
**Pull request by mrocklin**
This creates a vanilla Tornado HTTPServer on the Worker and Schedulers, and then optionally attaches a Bokeh application to that server.  

```python
In [1]: from dask.distributed import LocalCluster                                                        

In [2]: cluster = LocalCluster()                                                                         

In [3]: cluster.scheduler.http_server                                                                    
Out[3]: <tornado.httpserver.HTTPServer at 0x7f74fc66f4d0>

In [4]: cluster.scheduler.http_application.applications                                                  
Out[4]: [<bokeh.server.tornado.BokehTornado at 0x7f74ef9dd650>]
```

Fixes #3645 

The goal here is to allow us to always have an HTTP Server running, even if we aren't running the bokeh dashboard.  This would be helpful for example for prometheus and health checks, and to allow for other uses of the HTTP Server in the future.  I hope that it also makes it easier to configure HTTPS globally across the project.

However, while this cleans things up (at least in my opinion) this does introduce some possible breaking changes.  The HTTP Server is no longer a "service" (this concept never really caught on anyway) and I'm tempted to change various keywords from "dashboard" to "http" (although this could easily be reverted if we want more backwards compatibility).  Additionally I *think* I'm handling all of the Bokeh stuff correctly, but I'm working around considerable amount of engineering in that project.  

I originally cc'ed @jcrist and @jacobtomlinson on the original issue, but I think that both of them are out for life reasons.  cc'ing @quasiben @jrbourbeau and @TomAugspurger instead.

### TODOs:

- [x] Figure out how to handle dashboard keywords, and whether or not to rename them to http or something else
- [x] Make the different sets of HTTP configurable.  For example we probably always want the health checks, but may not always want the proxying or dashboard.
- [x] Add security back in
- [ ] ...

### CODE REVIEW

**Comment by mrocklin** [comment]
OK, for now I'm leaving the user-facing keywords as they were.  We can handle renaming things in a future PR if we think it makes sense.

**Comment by mrocklin** [comment]
The challenge with keeping the keywords as they were is that now we can have situations like the following:

    dask-scheduler --dashboard-address :8787 --no-dashboard

Which means "host the HTTP Server at port 8787, but don't include the Bokeh stuff".  This is a reasonable thing to do, but the spelling is unfortunate.  We might also consider the following:

    dask-scheduler --http-address :8787 --no-dashboard

And then we would deprecate `--dashboard-address` and route it to `--http-address` for a while.

**Comment by mrocklin** [comment]
@fjetter this may interest you all as well

**Comment by StephanErb** [comment]
Feature request: Please make it officially possible to add custom routes via a preload module. 

As of today the following code works great in a worker 
```py
@click.command()
@click.option("--my-custom-endpoint")
def dask_setup(instance, my_custom_endpoint):

    from distributed.dashboard.utils import RequestHandler
    from distributed.dashboard.worker import routes

    class MyHandler(RequestHandler):
        <snip>

    routes.append((my_custom_endpoint, MyHandler))
```
The code does not work for a scheduler though. For the scheduler the HTTP server instantiation and the preload does not align well, so one needs to dig deep into the internals:

```
    scheduler.services["dashboard"].server._tornado.add_handlers(
        r".*", [("/" + my_custom_endpoint, MyHandler)]
```

**Comment by TomAugspurger** [comment]
I don't think I'll have much useful feedback here. My only request is that we ensure we test against multiple versions of bokeh in our CI if we aren't already.

**Comment by mrocklin** [comment]
> Feature request: Please make it officially possible to add custom routes via a preload module.

Yes, I was also thinking about that.  My current thought is that we include a list of modules in configuration.  Maybe something like this:

```yaml
scheudler:
  http:
    route-modules:
    - distributed.http.core
    - distributed.http.scheduler.prometheus
    - distributed.http.scheduler.proxy
    - distributed.http.scheduler.json
```

We would expect these modules to have a specific structure, like having a `routes` variable.

Thoughts on how we could improve this @StephanErb ?

**Comment by mrocklin** [comment]
> I don't think I'll have much useful feedback here. My only request is that we ensure we test against multiple versions of bokeh in our CI if we aren't already.

I'm currently not assuming my changes in Bokeh.  Those changes will allow us to delete some code here in the future when our minimum version catch up.

**Comment by mrocklin** [comment]
Tests pass.  This is in a state where it could be merged.  I haven't added the ability to extend or remove sets of routes yet though.

**Comment by jcrist** [comment]
Apologies, I'm on leave right now without a computer (using my phone, not too easy to read code without a computer), this may not be relevant/already covered.

> We would expect these modules to have a specific structure, like having a routes variable.

Instead of assuming a standard module structure, I'd prefer the scheduler/worker objects have a set of public methods on them, and use the existing preload script mechanism to add routes/modify the server as needed. This seems more extensible/flexible to me than forcing the user to split their code up into modules with a specific structure. In many cases the routes module may also add plugins or additional services, so keeping http server routes as part of the existing extension points makes more sense to me.

```python
# pseudocode
def dask_setup(scheduler):
    scheduler.http_server.add_routes(...)
```

**Comment by mrocklin** [comment]
@jcrist did the reasoning above change your thinking at all?

**Comment by mrocklin** [comment]
OK, I've implemented the separate route-groups as modules approach.  The configuration looks like this:

```yaml
distributed:
scheduler:
    http:
      routes:
        - distributed.http.scheduler.prometheus
        - distributed.http.scheduler.info
        - distributed.http.scheduler.json
        - distributed.http.health
        - distributed.http.proxy
        - distributed.http.statics
  worker:
    http:
      routes:
        - distributed.http.worker.prometheus
        - distributed.http.health
        - distributed.http.statics
```

We look for a `routes` variable within the module.  

This doesn't preclude people from also using preload scripts.  I like it because it lets users turn off parts of our functionality (like the proxy, or info pages) in a way that is general enough to also extend to their own.

**Comment by mrocklin** [comment]
@StephanErb I encourage you to give this a try if you have the time and share any thoughts you may have.

**Comment by mrocklin** [comment]
Maybe @gforsyth has the background to be able to review here effectively?

**Comment by mrocklin** [comment]
This is at a point where I think that it is mergeable, and will have a positive impact.  If there isn't any pushback here I will go ahead and merge this in 48 hours.

File: distributed/http/scheduler/info.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        except (ImportError, StopIteration):
+            self.write({})
+
+
+class EventstreamHandler(WebSocketHandler):
+    def initialize(self, server=None, extra=None):
+        self.server = server
+        self.extra = extra or {}
+        self.plugin = WebsocketPlugin(self, server)
+        self.server.add_plugin(self.plugin)
```
[the remark this answers is not in the document]
Comment (gforsyth) [reply distributed/http/scheduler/info.py:169]: sorry, to clarify, this causes a connection reset when browsing to `/eventstream` and the traceback in the scheduler-window points to this line as the culprit

File: distributed/dashboard/tests/test_worker_bokeh.py [hunk 2]
```diff
@@ -47,37 +42,34 @@ def test_routes(c, s, a, b):
...
-@pytest.mark.skipif(
-    sys.version_info[0] == 2, reason="https://github.com/bokeh/bokeh/issues/5494"
-)
-@gen_cluster(client=True, worker_kwargs={"services": {("dashboard", 0): BokehWorker}})
+@gen_cluster(client=True, worker_kwargs={"dashboard": True})
 def test_simple(c, s, a, b):
-    assert s.workers[a.address].services == {"dashboard": a.services["dashboard"].port}
-    assert s.workers[b.address].services == {"dashboard": b.services["dashboard"].port}
+    # assert s.workers[a.address].services == {"dashboard": a.services["dashboard"].port}
+    # assert s.workers[b.address].services == {"dashboard": b.services["dashboard"].port}
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/dashboard/tests/test_worker_bokeh.py]: Thanks.  I've modified and uncommented these.

File: distributed/dashboard/tests/test_worker_bokeh.py [hunk 3]
```diff
@@ -47,37 +42,34 @@ def test_routes(c, s, a, b):
...
 
 
 @gen_cluster(
-    client=True, worker_kwargs={"services": {("dashboard", 0): (BokehWorker, {})}}
+    client=True, worker_kwargs={"dashboard": True},
 )
 def test_services_kwargs(c, s, a, b):
-    assert s.workers[a.address].services == {"dashboard": a.services["dashboard"].port}
-    assert isinstance(a.services["dashboard"], BokehWorker)
+    # assert s.workers[a.address].services == {"dashboard": a.services["dashboard"].port}
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/dashboard/tests/test_worker_bokeh.py]: Same.

File: distributed/dashboard/worker.py [hunk 4]
```diff
@@ -13,177 +6,28 @@
...
-                {"server": self.my_server, "extra": self.extra},
-            )
-            for url, cls in routes
-        ]
-
-        self.server._tornado.add_handlers(r".*", handlers)
+applications = {
+    "/status": status_doc,
+    "/counters": counters_doc,
+    "/crossfilter": crossfilter_doc,
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/dashboard/worker.py:29]: Hrm, this did cause me to realize that we weren't publishing the statics route to the workers (this is now fixed), but I am definitely able to access this route.  I'm not sure what's going on here.

File: distributed/http/scheduler/info.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/http/scheduler/info.py:169]: I've resolved this.  It looks like this class is currently untested.  I think that this is for @jacobtomlinson 's new "pew pew pew" plot (official name) which is currently a bit experimental.  

I've added a very minimal test, but hopefully it should catch dumb things like this.

File: distributed/http/scheduler/json.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        self.write(self.server.identity())
+
+
+class IndexJSON(RequestHandler):
+    def get(self):
+        with log_errors():
+            r = [url for url, _, _ in routes if url.endswith(".json")]
+            self.render(
+                "json-index.html", routes=r, title="Index of JSON routes", **self.extra
+            )
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/http/scheduler/json.py]: I've resolved this in a hacky way for now.  Long term though we should probably invest in an index page for everything.  I'll raise an issue.

File: distributed/http/tests/test_routing.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    response = await client.fetch("http://localhost:1234/three")
+    assert response.body.decode() == "one"
+
+    # from dask.distributed import LocalCluster
+    # async with LocalCluster(processes=False, asynchronous=True) as cluster:
+    #     dashboard = cluster.scheduler.services["dashboard"]
+    #     application.applications.append(dashboard.server._tornado)
+
+    #     response = await client.fetch("http://localhost:1234/status")
+    #     assert "bokeh" in response.body.decode()
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/http/tests/test_routing.py]: Done

File: distributed/http/scheduler/json.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/http/scheduler/json.py]: https://github.com/dask/distributed/issues/3677

**Comment by mrocklin** [comment]
Thank you for the careful review @gforsyth .  That was useful.

**Comment by mrocklin** [comment]
Merging this soonish.

**Pull request merged**